docs/references/http-gateway-protocol-spec.md

```diff
@@ -0,0 +1,471 @@
+	repeated string certified_request_headers = 1;
+	repeated string certified_query_parameters = 2;
+}
+
+message ResponseCertification {
+	oneof response_headers {
+    	ResponseHeaderList certified_response_headers = 1;
+    	ResponseHeaderList response_header_exclusions = 2;
+	}
+}
+
+message Certification {
+	oneof request {
+		RequestCertification request_certification = 1;
+		Empty no_request_certification = 2;
+	}
+	ResponseCertification response_certification = 3;
+}
+
+message ValidationArgs {
+	oneof certification {
+    	Certification certification = 1;
+    	Empty no_certification = 2;
+	}
+}
+```
+
+The syntax of the header is defined by the following [EBNF](https://en.wikipedia.org/wiki/Extended_Backus%E2%80%93Naur_form):
+
+```
+CHAR = /[^\0\n"]/
+STRING = '"', { CHAR }, '"'
+STRING-LIST = '[', { STRING }, ']'
+
+RESPONSE-HEADER-LIST = 'ResponseHeaderList{headers:',  STRING-LIST, '}'
+
+REQUEST-CERTIFICATION = 'RequestCertification{certified_request_headers:', STRING-LIST, ',certified_query_parameters:', STRING-LIST, '}'
+
+RESPONSE-CERTIFICATION = 'ResponseCertification{', ('response_header_exclusions:' | 'certified_response_headers:'), RESPONSE-HEADER-LIST, '}'
+
+CERTIFICATION = 'Certification{('no_request_certification: Empty{}' | 'request_certification:', REQUEST-CERTIFICATION), ',response_certification:', RESPONSE-CERTIFICATION, '}'
+
+VALIDATION-ARGS = 'ValidationArgs{', ('no_certification: Empty{}' | 'certification:', CERTIFICATION), '}'
+
+HEADER-VALUE = 'default_certification(', VALIDATION-ARGS, ')'
+
+HEADER = 'IC-CertificateExpression: ', HEADER-VALUE
+```
+
+### Request Hash Calculation
+
+The request hash is calculated as follows:
+
+1. Let `request_headers_hash` be the [representation-independent hash](https://internetcomputer.org/docs/current/references/ic-interface-spec#hash-of-map) of the request headers:
+   - Only include headers listed in the `certified_request_headers` field of [the certificate expression header](#the-certificate-expression-header).
+     - If the field is empty or no value was supplied, no headers are included.
+     - Headers can be repeated and each repetition should be included.
+   - Include an additional `:ic-cert-method` header that contains the HTTP method of the request.
+   - Include an additional `:ic-cert-query` header that contains a value according to the following steps:
+     - Parse the query string and build a list of tuples `(<query_param_name>, <query_param_value>)` while maintaining the order.
+     - Exclude all tuples where `<query_param_name>` does not exactly match a value listed in the `certified_query_parameters` field of [the certificate expression header](#the-certificate-expression-header). If `certified_query_parameters` is empty then the resulting list of tuples should also be empty.
+     - Concatenate each `<query_param_name>` with the corresponding `<query_param_value>` and then concatenate all of these concatenations using the original separators and order.
```

Comment:
Why does the order matter?

Reply:
It's to make sure that any endpoints that happen to rely on the order can have different hashes for different orders. I'd generally consider it to be bad practice to rely on the order, but the idea was to be flexible in case someone wanted to do that for some reason...

Reply:
Ok, got it. Can be resolved, I just don't have the rights to resolve it myself.